src/Peachpie.CodeAnalysis/FlowAnalysis/AnalysisFacts.cs

```diff
@@ -522,5 +523,52 @@ private static bool TryGetVariableHandle(IVariableReference boundvar, FlowState
 
             return isFeasible;
         }
+
+        /// <summary>
+        /// Returns whether the given type can be used as an array key.
+        /// </summary>
+        public static bool IsValidKeyType(IBoundTypeRef type)
+        {
+            if (type is BoundPrimitiveTypeRef pt)
+            {
+                switch (pt.TypeCode)
+                {
+                    case PhpTypeCode.Boolean:
+                    case PhpTypeCode.Long:
+                    case PhpTypeCode.Double:
+                    case PhpTypeCode.String:
+                    case PhpTypeCode.WritableString:
+                    case PhpTypeCode.Null:
+                        return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// If present, transforms the given constant value to a string corresponding to the key under which the item is stored in an array.
+        /// </summary>
+        /// <param name="keyConst">Constant value of the key.</param>
+        /// <param name="keyString">If <paramref name="keyConst"/> contains a value, the key as a string.</param>
+        /// <returns>Whether the value was constant at all.</returns>
+        public static bool TryGetCanonicKeyStringConstant(Optional<object> keyConst, out string keyString)
```

Comment:
this causes a lot of string allocations :( 

the conversion to array key is described in runtime by struct `IntStringKey`. Can we do the similar?

return Tuple<string, long>
convert all numbers to long
convert string to long if possible, otherwise keep the string
else return false

this would avoid repetitious conversions of number -> string (which is the most common case) and would speedup the duplicity check

Reply:
Good idea, fixed in 2eee6f3ae66e8a265d3704329c9789ab05bc34bd.